README.md
node-credit-cards
=================

Validate & extract metadata from credit cards.

=== Basic usage

```js
var creditCards = require('credit-cards')

creditCards.isValid("4024 0071 2016 8322")
=> true
creditCards.isValid("4034 0071 2016 8322")
=> false

creditCards.getTypeCode("4034 0071 2016 8322")
=> 'VI'
creditCards.getTypeName("4034 0071 2016 8322")
=> 'Visa'

creditCards.getInfo('3478 7964 2068 802')
=> { valid: true, code: 'AX', name: 'American Express' }
```

=== Advanced usage

```js
var creditCards = require('credit-cards')

var scoped = creditCards({ accepted: ['VI', 'AX'] })

scoped.isValid("4024 0071 2016 8322")
=> true
scoped.getTypeCode("4034 0071 2016 8322")
=> 'VI'
scoped.isAccepted('5290 6473 7075 0487')
=> false
scoped.getInfo('5290 6473 7075 0487')
=> { valid: true, code: 'CA', name: 'Master Card', accepted: false }
```


CHANGELOG.md
### v1.0.1
- Fix missing lib/index.js

### v1.0.0
- Convert to TypeScript (same API, package compiled to JS). No breaking changes.


test/lib/index_spec.js
var should      = require('should');
var creditCards = require('../../lib/');

describe("Credit cards", function () {
  describe("base functionality", function () {
    describe("#isValid", function () {
      it("should return true for valid numbers", function () {
         should.strictEqual(creditCards.isValid(" 402 400712016 8322 "), true);
      });

      it("should return false for invalid numbers", function () {
         should.strictEqual(creditCards.isValid(" 4034007 12016832 2"), false);
      });

      it("should return false for test numbers", function () {
         should.strictEqual(creditCards.isValid("4111111111111111"), false);
      });
    });

    describe("#getInfo", function () {
      it("should return correct data", function () {
        creditCards.getInfo(' 3478 7964 2068 802').should.eql({
          valid: true, code: 'AX', name: 'American Express', test: false
        });
      });
      it("should not try to accept non-existant caards", function () {
        should.not.exist(creditCards.getInfo(' 0000 0000 0000 0000', {accepted:['none']}).accepted);
      });
    });
  });

  describe("fancy functionality", function () {
    var fancyCards;

    before(function () {
      fancyCards = creditCards({ accepted: ['VI', 'AX'] });
    });

    describe("#isValid", function () {
      it("should return true for valid numbers", function () {
         should.strictEqual(fancyCards.isValid(" 40240 07120168322"), true);
      });

      it("should return false for invalid numbers", function () {
         should.strictEqual(fancyCards.isValid("4034007120 168322"), false);
      });
    });

    describe("#isAccepted", function () {
      it("should return false for not accepted cards", function () {
        should.strictEqual(fancyCards.isAccepted('529064 73707 50487 '), false);
      });

      it("should return true for accepted cards", function () {
        should.strictEqual(fancyCards.isAccepted(' 343018 837748043'), true);
      });
    });

    describe("#getInfo", function () {
      it("should return correct data", function () {
        fancyCards.getInfo('3478 79642068802').should.eql({
          valid: true, code: 'AX', name: 'American Express', test: false, accepted: true
        });

        fancyCards.getInfo(' 52906 473707 50487').should.eql({
          valid: true, code: 'CA', name: 'Master Card', test: false, accepted: false
        });
      });
    });
  });
});


test/src/index_test.ts
const should      = require('should');
const creditCards = require('../../src/');

describe("Credit cards", () => {
  describe("base functionality", () => {
    describe("#isValid", () => {
      it("should return true for valid numbers", () => {
         should.strictEqual(creditCards.isValid(" 402 400712016 8322 "), true);
      });

      it("should return false for invalid numbers", () => {
         should.strictEqual(creditCards.isValid(" 4034007 12016832 2"), false);
      });

      it("should return false for test numbers", () => {
         should.strictEqual(creditCards.isValid("4111111111111111"), false);
      });
    });

    describe("#getInfo", () => {
      it("should return correct data", () => {
        creditCards.getInfo(' 3478 7964 2068 802').should.eql({
          valid: true, code: 'AX', name: 'American Express', test: false
        });
      });
      it("should not try to accept non-existant caards", () => {
        should.not.exist(creditCards.getInfo(' 0000 0000 0000 0000', { accepted: ['none'] }).accepted);
      });
    });
  });

  describe("Advanced functionality", () => {
    let advanced;

    before(() => {
      advanced = creditCards({ accepted: ['VI', 'AX'] });
    });

    describe("#isValid", () => {
      it("should return true for valid numbers", () => {
         should.strictEqual(advanced.isValid(" 40240 07120168322"), true);
      });

      it("should return false for invalid numbers", () => {
         should.strictEqual(advanced.isValid("4034007120 168322"), false);
      });
    });

    describe("#isAccepted", () => {
      it("should return false for not accepted cards", () => {
        should.strictEqual(advanced.isAccepted('529064 73707 50487 '), false);
      });

      it("should return true for accepted cards", () => {
        should.strictEqual(advanced.isAccepted(' 343018 837748043'), true);
      });
    });

    describe("#getInfo", () => {
      it("should return correct data", () => {
        advanced.getInfo('3478 79642068802').should.eql({
          valid: true, code: 'AX', name: 'American Express', test: false, accepted: true
        });

        advanced.getInfo(' 52906 473707 50487').should.eql({
          valid: true, code: 'CA', name: 'Master Card', test: false, accepted: false
        });
      });
    });
  });
});


src/index.ts
const luhn = require('luhn');

import { types } from './types';

import type {
  CreditCardCode,
  CreditCardName,
  CreditCardType,
  CreditCardInfo,
  IsAcceptedOptions,
} from './types';

const allCodes = types.map((type) => type.code);

const testNumbers = [
  '378282246310005',  '371449635398431',  '378734493671000','30569309025904', '38520000023237',
  '4111111111111111', '4012888888881881', '4222222222222',
  '5555555555554444', '5105105105105100 ',
  '6011111111111117', '6011000990139424',
];

const clean = (number) => number.replace(/\s+/g, '');

const isTestNumber = (number) => testNumbers.includes(clean(number));

const isValid = (number: string) => {
  number = clean(number);

  if (luhn.validate(number)) {
    return !isTestNumber(number);
  } else {
    return false;
  }
}

const getType = (number: string) => {
  let card;

  for (let i = 0; i < types.length; i++) {
    card = types[i];

    if (card.pattern.test(clean(number))) {
      return card;
    }
  }
}

const getTypeCode = (number: string) => {
  var type = getType(clean(number));

  if (type) {
    return type.code;
  }
}

const getTypeName = (number: string) => {
  var type = getType(clean(number));

  if (type) {
    return type.name;
  }
}

const isAccepted = (number: string, opts: IsAcceptedOptions) => {
  number = clean(number);
  return isValid(number) && opts.accepted.includes(getTypeCode(number));
};

const getInfo = (number: string, opts?: IsAcceptedOptions): CreditCardInfo => {
  number = clean(number);

  const info: CreditCardInfo = { valid: isValid(number) };
  const card = getType(number);

  if (card) {
    info.code = card.code;
    info.name = card.name;
    info.test = isTestNumber(number);

    if (opts?.accepted) {
      info.accepted = isAccepted(number, opts);
    }
  }

  return info;
}

const functions = {
  isTestNumber,
  isValid,
  getTypeName,
  getTypeCode,
  getInfo,
  allCodes,
};

type AdvancedHelper = typeof functions & {
  isAccepted(number: string): boolean;
  getInfo(number: string): CreditCardInfo;
};

function advancedHelper (opts: IsAcceptedOptions): AdvancedHelper {
  function F () {
    this.isAccepted = (number: string) => isAccepted(number, opts);
    this.getInfo    = (number: string) => getInfo(number, opts);
  };

  F.prototype = functions;

  return new F;
}

export = Object.assign(
  advancedHelper,
  functions,
);


src/types.ts
export const types = [
  { code: 'VI', name: 'Visa',                pattern: /^4[0-9]{12}(?:[0-9]{3})?$/        },
  { code: 'CA', name: 'Master Card',         pattern: /^5[1-5][0-9]{14}$/                },
  { code: 'DC', name: 'Diners Club',         pattern: /^3(?:0[0-5]|[68][0-9])[0-9]{11}$/ },
  { code: 'AX', name: 'American Express',    pattern: /^3[47][0-9]{13}$/                 },
  { code: 'DS', name: 'Discover',            pattern: /^6(?:011|5[0-9]{2})[0-9]{12}$/    },
  { code: 'TP', name: 'UAPT',                pattern: /^1[0-9]{14,15}$/                  },
  { code: 'JC', name: 'Japan Credit Bureau', pattern: /^(?:2131|1800|35\d{3})\d{11}$/    },
] as const;

export type CreditCardCode = typeof types[number]['code'];
export type CreditCardName = typeof types[number]['name'];
export type CreditCardType = typeof types[number];

export type CreditCardInfo = {
  valid    : boolean;
  code?    : CreditCardCode;
  name?    : CreditCardName;
  test?    : boolean;
  accepted?: boolean;
};

export type IsAcceptedOptions = {
  accepted: CreditCardCode[];
};


lib/index.js
var luhn = require('luhn').luhn;

var types = [
  {code: 'VI', name: 'Visa',                pattern: /^4[0-9]{12}(?:[0-9]{3})?$/        },
  {code: 'CA', name: 'Master Card',         pattern: /^5[1-5][0-9]{14}$/                },
  {code: 'DC', name: 'Diners Club',         pattern: /^3(?:0[0-5]|[68][0-9])[0-9]{11}$/ },
  {code: 'AX', name: 'American Express',    pattern: /^3[47][0-9]{13}$/                 },
  {code: 'DS', name: 'Discover',            pattern: /^6(?:011|5[0-9]{2})[0-9]{12}$/    },
  {code: 'TP', name: 'UAPT',                pattern: /^1[0-9]{14,15}$/                  },
  {code: 'JC', name: 'Japan Credit Bureau', pattern: /^(?:2131|1800|35\d{3})\d{11}$/    }
];
var allCodes = types.map(function (type) { return type.code; });

var testNumbers = [
  '378282246310005',  '371449635398431',  '378734493671000','30569309025904', '38520000023237',
  '4111111111111111', '4012888888881881', '4222222222222',
  '5555555555554444', '5105105105105100 ',
  '6011111111111117', '6011000990139424'
];

function clean (number) {
  return number.replace(/\s+/g, '');
}

function isTestNumber (number) {
  return testNumbers.indexOf(clean(number)) != -1;
}

function isValid (number) {
  number = clean(number);

  if (luhn.validate(number)) {
    return !isTestNumber(number);
  } else {
    return false;
  }
}

function getType (number) {
  var i, card;

  for (i = 0; i < types.length; i++) {
    card = types[i];

    if (card.pattern.test(clean(number))) return card;
  }
}

function getTypeCode (number) {
  var type = getType(clean(number));

  if (type) return type.code;
}

function getTypeName (number) {
  var type = getType(clean(number));

  if (type) return type.name;
}

function isAccepted (number, opts) {
  number = clean(number);
  return isValid(number) && opts.accepted.indexOf(getTypeCode(number)) != -1;
}

function getInfo (number, opts) {
  var card;
  var info = {};
  number = clean(number)

  info.valid = isValid(number);

  if (card = getType(number)) {
    info.code = card.code;
    info.name = card.name;
    info.test = isTestNumber(number);

    if (opts && 'accepted' in opts) {
      info.accepted = info.valid && opts.accepted.indexOf(card.code) != -1;
    }
  }

  return info;
}

function fancyCreditCards (opts) {
  if (typeof opts != 'object') opts = {};
  if (!('accepted' in opts)) opts.accepted = allCodes;

  function F () {
    this.isAccepted = function (number) {
      return isAccepted(number, opts);
    };
    this.getInfo    = function (number) {
      return getInfo(number, opts);
    };
  };

  F.prototype = module.exports;

  return new F;
}

module.exports = fancyCreditCards;
module.exports.isTestNumber = isTestNumber;
module.exports.isValid      = isValid;
module.exports.getTypeName  = getTypeName;
module.exports.getTypeCode  = getTypeCode;
module.exports.getInfo      = getInfo;
